### ISSUE TITLE
Handle multiple connections to the same peer

### ISSUE THREAD

**Issue by arnetheduck** [status-im/nim-libp2p#236]
https://github.com/status-im/nim-beacon-chain/pull/1204 installs a filter so that multiple concurrent connection attempts to the same peer are prevented - this should be handled at a libp2p level, making sure that all such attempts are properly signalled when one of the connections succeeds (ie they all get connected to the same "connection", that they all get satisfied if an incoming connection appears that "connects" the peer etc.

**Comment by sinkingsugar** [comment]
Hmm, I tend to disagree tho.. when I use any API/SDK library such cases fall into bad use of the library (e.g. calling stuff multiple times and/or concurrently).
I really don't feel like this should be libp2p responsibility.. libp2p is supposed to be a low level and lightweight networking layer after all.

Well that is my personal take, happy to discuss and change idea.. but I just feel you are biased by NBC a bit.

**Comment by dryajov** [comment]
This is actually not as straightforward as it might appear at first, here is some context

TL;DR

- We can control that for outgoing connections
- No way to control it for incoming connections at the libp2p level, as there isn't a way of communicating to the remote peer that it already has a connection established
- For a better experience, we potentially need peer scoring and connection statistics

--------------------------

Outgoing dials should be synchronized and a peer shouldn't dial the same remote many times concurrently, ie, internally there should be a lock that prevents that.

Incoming connections are more complicated, because we can't "synchronize" anything on the remote peer, thus we can't prevent it from dialing us many times. So the issue is what to do when we already have a connection to the peer, either outgoing or incoming and we're getting yet another incoming connection? 

We could:

a) Drop the connection and move on. But that is problematic, because there is no way of communicating that the connection is duplicated and to the remote peer it will look just as the dial failed, at that point the peer might or might not retry. This can dramatically decrease the mesh connectivity and our implementation's interoperability.

b) Accumulate the duplicate incoming connections and allow the remote peer to continue as if nothing happened, perhaps decreasing the peer's score (if the mechanism is in place) by some amount, this will handle spam and buggy clients in a more elegant manner.

Most clients implement "b" in some form. My approach is:

- Outgoing dials are locked, so no two outgoing connections should exist for the same peer
- Incoming connections are accumulated and the peer is allowed to open many incoming connections. Once there is some way of scoring peers, then we'll be able to throttle/disconnect the peer. To prevent this from being an attack vector, there will be a hard upper limit for incoming connections per peer (3-5?), that once reached will drop that peer.
- When the local peer tries to dial a remote peer for which there is already a connection, the existing connection should be returned. If there are multiple connections to the same peer, there should be some algorithm for choosing the best one. However, in order to be able to decide which connection is best, metrics/stats such as throughput, latency, etc, need to be collected for the connection. Unfortunately, we don't yet collect connection stats, so, for now we're stuck with something much simpler. IMO, the best metric we have right now is weather the connection is outgoing or incoming, incoming connections should be preferred as we're in control of their lifecycle, thus, the selection algorithm will pick an outgoing connection if it exist and if not, then it will simply pick the first outgoing connection from the list.
